src/platform/qt/library/LibraryController.cpp

```diff
@@ -0,0 +1,173 @@
+/* Copyright (c) 2014-2017 waddlesplash
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+#include "LibraryController.h"
+
+#include "../GBAApp.h"
+#include "LibraryGrid.h"
+#include "LibraryTree.h"
+
+namespace QGBA {
+
+void AbstractGameList::addGames(QList<LibraryEntry*> items) {
+	for (LibraryEntry* o : items) {
+		addGame(o);
+	}
+}
+void AbstractGameList::removeGames(QList<LibraryEntry*> items) {
+	for (LibraryEntry* o : items) {
+		removeGame(o);
+	}
+}
+
+LibraryController::LibraryController(QWidget* parent_widget, const QString& path, ConfigController* config)
+	: QStackedWidget(parent_widget),
+	  m_config(config)
+{
+	if (!path.isNull()) {
+		m_library = mLibraryLoad(path.toUtf8().constData());
+	} else {
+		m_library = mLibraryCreateEmpty();
+	}
+
+	mLibraryAttachGameDB(m_library, GBAApp::app()->gameDB());
+
+	m_libraryTree = new LibraryTree(this);
+	addWidget(m_libraryTree);
+	connect(m_libraryTree, &LibraryTree::StartGame, this, &LibraryController::startGame);
+
+	m_libraryGrid = new LibraryGrid(this);
+	addWidget(m_libraryGrid);
+	connect(m_libraryGrid, &LibraryGrid::StartGame, this, &LibraryController::startGame);
+
+	refresh();
+	selectLastBootedGame(); // TODO: doesn't belong here
+
+	setViewStyle(LibraryStyle::STYLE_TREE);
+}
+
+LibraryController::~LibraryController() {
+	for (LibraryEntry* file : m_games.values()) {
+		delete file;
```

Comment:
Hmm. Don't think QList has a pointer getter, and there need to be uniquely identifiable pointers. Will see what I can do.